Add a script for auto-generating dependency summaries.

rfk opened the pull request:
A little experiment in automatically generating a list of dependencies for each of our built artifacts.

This does a similar thing to [`cargo lichking bundle`](https://github.com/Nemo157/cargo-lichking/) but with a few key differences:

* Thanks to `cargo tree`, it only traverses dependencies of a specific package build, rather than pulling in dependencies based on the union of features from all packages in the workspace. In particular, it doesn't think that hyper is a dependency of the fenix megazord.

* It attempts to fetch license text from the remote github repo if it's not available in the local package cache (which in practice only seems to be required for a small number of dependencies).

If we like the result here, perhaps we could extending or forking lichking to include the above functionality rather than maintaining it as a little python script, but this was the simplest way for me to experiment with the possibilities here.

@thomcc could you please try this out and see if it matches your expectations re: what dependencies are or are not included for Fenix?

(For this to be useful in practice, I think we would need to manually audit that it is doing the right thing, and manually audit any changes when adding or updating dependencies, but this seems like an OK place to start).

thomcc commented:
Two quick comments:
1. We'll probably need to pass `--target` into cargo tree (we currently gate certain deps on target, and would like to gate more, since it works more predictably / reliably than --feature)
2. I think we might want to do some form of deduping of license info, to avoid e.g. each tokio crate bringing in its own copy of its licenses, as well as all the duplicate copies of apache2.

rfk commented:
> as well as all the duplicate copies of apache2.

I liked the grouping that you did for Firefox for iOS in this regard, it seems worth taking the same approach here.

thomcc commented:
For completeness (I think your script is better, and at a minimum, I'm happy I didn't have to write it 😅), my script is https://gist.github.com/thomcc/8523cca446ab632f2d48b23893fbe74e. I wrote it while coming down with a cold, and the code is both bad, and some of the logic (not including build dependencies) is dubious.

I think using cargo tree as a starting point is clever. We also might want to use `cargo +nightly build -Z unstable-options --build-plan` which includes the *true* dependency tree as part of its output (in the `inputs` property, IIRC).

rfk commented:
So, a brief update here.

I've changed this to use `cargo build --build-plan` as suggested, after confirming that it produced equivalent results to `cargo tree` (the only discrepency I noticed was the `cargo tree` always included `winapi` in the dependencies, which seemed like a false positive).  I've also added `--target` specifiers and am now pretty confident that it's identifying the correct subset of dependencies for each build artifact.

I've also added a bunch of explicit overrides to fill in gaps in the available cargo package metadata, so that nothing here needs to guess about the status of a dependency - either we can unambiguously determine its details from `cargo metadata` output, or we have manually checked it and recorded the results in this file.

Finally, I've got a little experiment going where I generated the listings for a bunch of our existing artifacts, committed them to the repo, and set up circle to error out if it can't regenerate exactly the same file. Similar to our use of `cargo fmt -- --check`, this should help guard against any unexpectedchanges in our dependency tree, say from third-party dependency changes with new licenses that we haven't audited yet. It replaces the existing license-checking CI job.

I'll poke at the formatting a bit more, but I think this is getting close to graduating from "draft".

rfk commented:
> Finally, I've got a little experiment going where I generated the listings for a bunch of our existing 
> artifacts, committed them to the repo, and set up circle to error out if it can't regenerate exactly
> the same file.

This definitely works, because I had to boot up my Mac to generate the correct dependency list for the iOS megazord in order to get this passing CI!

rfk commented on tools/dependency_summary.py:
```diff
@@ -0,0 +1,737 @@
+            'cargo', '+nightly', '-Z', 'unstable-options', 'build',
+            '--build-plan',
+            '--quiet',
+            '--frozen',
```
Note to self: `--locked` is probably sufficient here.

rfk commented:
I think I'm ready to call this "ready for review", and we can tweak in followups as needed.

thomcc commented on tools/dependency_summary.py:
```diff
@@ -0,0 +1,815 @@
+# We list both the expected value from `cargo metadata` and the replacement value, to
+# guard against accidentally overwriting future upstream changes in the metadata.
+#
+# Let's try not to add any more dependencies that require us to edit this list!
```
I'd feel better about this if it didn't include test/examples/benchmarks/etc.

rfk replied:
I've succeeded in removing many of the test/example-only dependencies from the list, FWIW.

thomcc commented on tools/dependency_summary.py:
```diff
@@ -0,0 +1,815 @@
+            "fixup": "https://raw.githubusercontent.com/bryant/argon2rs/master/LICENSE",
+        }
+    },
+    "cloudabi": {
```
This could also go in the list of 'target specific crates for targets we do not support'

thomcc commented on tools/regenerate_dependency_summaries.sh:
```diff
@@ -0,0 +1,7 @@
+python3 ./tools/dependency_summary.py > ./DEPENDENCIES.md
+python3 ./tools/dependency_summary.py --all-android-targets --package lockbox > megazords/lockbox/DEPENDENCIES.md
+python3 ./tools/dependency_summary.py --all-android-targets --package fenix > megazords/fenix/DEPENDENCIES.md
+python3 ./tools/dependency_summary.py --all-android-targets --package reference-browser > megazords/reference-browser/DEPENDENCIES.md
```
Hmm, now that #1103 renames `reference-browser-megazord` to `full-megazord` I guess we'll probably need to track down whoever is including those us transitively to make sure they have this too.

(In that sense, maybe it would have been better for that PR to break them so that they reach out to us :p)

rfk commented:
By way of a quick update, I've changed this to not include test or example dependencies, and to fix the misunderstanding about `--all-targets` meaning "all known platform targets" rather than "all defined build targets". However, I think we should wait for #1103 to land before looking at landing this.

rfk commented:
> However, I think we should wait for #1103 to land before looking at landing this.

This landed, and I've rebased the work here on top of it. Having the full-megazord available has lead to some nice simplifications, because there an obvious small set of targets that will pull in all possible dependencies (the combination of "megazord" and "megazord_ios").

I'd like to land this so we can iterate on some of the details in smaller PRs, such as automatically including a dependency summary alongside our various built artifacts. @thomcc could you please take another look? @eoger I'd appreciate your feedback as well.

One possible downside of this approach, is that PRs from dependabot that introduce new transitive dependencies, may start to fail CI in the "check dependencie" step. That's on purpose, but it may prove to be a giant pain in practice.

eoger commented on tools/dependency_summary.py:
```diff
@@ -0,0 +1,844 @@
+    "adler32": ["ext-zlib"],
+    "nss_sys": ["ext-nss", "ext-nspr"],
+    "openssl-sys": ["ext-openssl"],
+    "rc_crypto": ["ext-ring"],
```
Maybe add a comment here that explains the situation?

rfk commented on docs/dependency-management.md:
```diff
@@ -47,28 +49,42 @@ We try to balance this responsibility against the many benefits of using existin
+
+* [OpenSSL](https://www.openssl.org/source/)
+* [SQLCipher](https://github.com/sqlcipher/sqlcipher)
+* [Protobuf](https://github.com/protocolbuffers/protobuf)
```
@rfk need to add NSS to this list.

thomcc commented on tools/dependency_summary.py:
```diff
@@ -0,0 +1,844 @@
+            "fixup": "COPYING",
+        }
+    },
+    # These packages do not include their license file in their release distributions,
```
It's unfortunate we need this, but yeah, I don't see an alternative. Seems like it might add a lot of review churn, unfortunately.
